### PR TITLE
Performance optimization by avoiding unnecessary passes

### PR DESCRIPTION
**Pull request by darksylinc**
# 🦟 Bug fix

No ticket created

## Summary

Ignition is pointlessly performing two passes when one is enough, as the
depth buffer data is already filled.

In the case of `Ogre2DepthCamera` we cannot do the same with
colorTexture & depthTexture though; because the way they are filled is different (colorTexture contains particles
and potentially other stuff that might write to depth; depthTexture doesn't)

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: ogre2/src/Ogre2DepthCamera.cc [hunk 1]
```diff
@@ -690,12 +681,15 @@ void Ogre2DepthCamera::CreateDepthTexture()
     particleDepthTexDef->fsaa = "0";
     particleDepthTexDef->textureFlags &= ~Ogre::TextureFlags::Uav;
 
-    Ogre::RenderTargetViewDef *rtvparticleDepthTex =
-      baseNodeDef->addRenderTextureView("particleDepthTexture");
-    rtvparticleDepthTex->setForTextureDefinition(
-      "particleDepthTexture", particleDepthTexDef);
+    // Auto setup the RTV then manually override the depth buffer so
+    // it uses the one we created (and thus we can sample from it later)
```
Comment (iche033) [inline ogre2/src/Ogre2DepthCamera.cc:685]: do we need to change the `depthBufferId` to something other than `Ogre::DepthBuffer::POOL_NON_SHAREABLE;`?

**Review by iche033 — APPROVED** [review]
nice, works for me. Verified that the particles in depth camera look good and also mouse camera control is also working as expected. Just have one minor question.

File: ogre2/src/Ogre2DepthCamera.cc (hunk 1 above)
Comment (darksylinc) [reply ogre2/src/Ogre2DepthCamera.cc:685]: Hi!

This change from 2.1 to 2.2+

Now depthBufferId should be getting ignored when the texture is `PFG_D32_FLOAT` (or any other depth format) and also ignored when the RTV is specifically linked to a depth buffer.

IIRC `POOL_NON_SHAREABLE` should be avoided by user code now. It is still used internally, but makes no sense in user code.

In 2.1 when you created a depth texture, you were not actually creating a depth texture. You set the depth pool ID to access the contents of an already created depth buffer associated with the render target with the same parameters and pool ID.

With POOL_NON_SHAREABLE you would prevent its contents from later being overridden by a later pass.

In 2.2 when you create a depth texture, you're actually creating a unique depth texture. And use RTVs to associate that depth texture with a colour target for rendering in a specific pass (and thus now depth pools are ignored).

The pool ID system only remains for throwable depth buffers for which you don't want to access/sample its depth contents and are automatically created and reused.

Comment (iche033) [reply ogre2/src/Ogre2DepthCamera.cc:685]: got it, thanks for the explanation!

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-01-citadel-edifice-fortress/1313/1

**Pull request merged**